cobalt/media/player/web_media_player_impl.cc

```diff
@@ -794,6 +809,83 @@ void WebMediaPlayerImpl::OnDemuxerOpened() {
   GetClient()->SourceOpened(chunk_demuxer_.get());
 }
 
+void WebMediaPlayerImpl::CheckDemuxerUnderflow() {
+  DCHECK_EQ(main_loop_, base::MessageLoop::current());
+  DCHECK(pipeline_);
+  DCHECK(!state_.is_media_source || chunk_demuxer_);
+  DCHECK(!state_.is_progressive || progressive_demuxer_);
+
+  if (state_.is_url_based) {
+    return;
+  }
+
+  if ((state_.is_media_source && chunk_demuxer_->GetIsEndOfStreamReceived()) ||
+      (state_.is_progressive &&
+       progressive_demuxer_->GetIsEndOfStreamReceived())) {
+    // End of stream has been received, so we don't check underflow.
+    SetDemuxerUnderflow(false);
+    return;
+  }
+
+  constexpr base::TimeDelta kDemuxerUnderflowThreshold =
+      base::TimeDelta::FromMilliseconds(100);
+  constexpr base::TimeDelta kEndOfStreamEpsilon =
+      base::TimeDelta::FromSeconds(1);
+
+  base::TimeDelta media_time = pipeline_->GetMediaTime();
+  base::TimeDelta duration = pipeline_->GetMediaDuration();
+  if (duration != ::media::kInfiniteDuration &&
+      media_time + kEndOfStreamEpsilon >= duration) {
+    // We don't check underflow at the end of the stream.
+    SetDemuxerUnderflow(false);
+    return;
+  }
+
+  ::media::Ranges<base::TimeDelta> buffered_ranges =
+      pipeline_->GetBufferedTimeRanges();
+  base::TimeDelta buffered_duration;
+  for (size_t i = 0; i < buffered_ranges.size(); i++) {
+    if (buffered_ranges.end(i) < media_time) {
+      continue;
+    }
+    if (buffered_ranges.start(i) > media_time) {
+      break;
+    }
+
+    buffered_duration = buffered_ranges.end(i) - media_time;
+  }
+
+  if (buffered_duration > kDemuxerUnderflowThreshold) {
+    // We have enough buffered data.
+    SetDemuxerUnderflow(false);
+    demuxer_underflow_timer_.Start(
+        FROM_HERE, buffered_duration - kDemuxerUnderflowThreshold, this,
```

Comment:
1. Consider adjusting for case of playback rate > 1.0 here.
2. This callback happens on the main thread, which is often blocked for hundreds of milliseconds, so we may want to take that into consideration.
3. Consider adding JS based tests for enough coverage of different situations, maybe with YTS.

Reply:
1. Adjusted the duration for playback rate.
2. I think it's ok if the callback is delayed. The worst case is the demuxer has underflow and the player doesn't notice of it, which is same as the existing implementation.
3. Yes, agree, b/319325203 is filed to trace it.

Reply:
Let's put a comment for situation two.